internal/kubernetes/util.go

```
@@ -417,6 +418,184 @@ func GetAnnotationFromPodOrController(annotationKey string, pod *core.Pod, store
+			return false
+		}
+	}
+	return true
+}
+
+// PruneErrors return a new search result with all errors removed. If handler functions for error are provided, they are called for each error that is pruned
+// The boolean that is return is true is No error was found and pruned.
+// The receiver is not modified.
+func (a *AnnotationSearchResult[T]) PruneErrors(nodeErrFunc func(node *core.Node, err error), podErrFunc func(pod *core.Pod, err error)) (AnnotationSearchResult[T], bool) {
+	noError := true
+	result := AnnotationSearchResult[T]{
+		Found: a.Found,
+	}
+	if a.errorConv != nil {
+		noError = false
+		if nodeErrFunc != nil && a.Node != nil {
+			nodeErrFunc(a.Node, a.errorConv)
+		}
+	} else {
+		result.Found = true
+		result.Node = a.Node
+		result.Value = a.Value
+	}
+
+	if a.PodResults == nil {
+		return result, noError
+	}
+	result.PodResults = map[string][]AnnotationSearchResultOnPod[T]{}
+	for k, pr := range result.PodResults {
+		for _, v := range pr {
+			if v.errorConv != nil {
+				noError = false
+				if podErrFunc != nil && v.Pod != nil {
+					podErrFunc(v.Pod, v.errorConv)
+				}
+			} else {
+				result.Found = true
+				result.PodResults[k] = append(result.PodResults[k], v)
+			}
+		}
+
+	}
+	if len(result.PodResults) == 0 {
+		result.PodResults = nil
+	}
+	return result, noError
+}
+
+func GetAnnotationFromNodeAndThenPodOrController[T comparable](ctx context.Context, podIndexer index.PodIndexer, store RuntimeObjectStore, converter func(string) (T, error), annotationKey string, node *core.Node, stopIfFoundOnNode, stopIfFoundOnPod bool) (AnnotationSearchResult[T], error) {
+	var result AnnotationSearchResult[T]
+	if node.Annotations != nil {
+		if valueStr, ok := node.Annotations[annotationKey]; ok {
+			result.Value, result.errorConv = converter(valueStr)
+			result.Found = (result.errorConv == nil)
+			if stopIfFoundOnNode {
+				return result, nil
+			}
+		}
+	}
+	if podIndexer == nil {
+		return result, nil
+	}
```

I think we should return an error in this case, as the result could be interpreted as "there is no annotation anywhere".

---

agree, fixed.